e2e/e2e_test.go

```diff
@@ -0,0 +1,301 @@
+		tt := tt
+		t.Run(tt.desc, func(t *testing.T) {
+			l := log.New(os.Stdout, fmt.Sprintf("[%s] ", t.Name()), 0)
+			dnsName := toDNSName(t.Name())
+
+			wd, err := os.Getwd()
+			if err != nil {
+				t.Fatalf("failed to get working directory: %s", err)
+			}
+			kubeConfFile := path.Join(wd, "kubeconfig-e2e."+dnsName)
+
+			// Delete old kubeconfig
+			if err := os.Remove(kubeConfFile); err != nil && !os.IsNotExist(err) {
+				t.Fatalf("failed to delete kubeconfig %q: %s", kubeConfFile, err)
+			}
+
+			// Create space.
+			storeName := toS3Name(fmt.Sprintf("%s-%s", os.Getenv(kopsEnvVarClusterName), dnsName))
+			if err := s3Cl.deleteSpace(storeName); err != nil {
+				t.Fatalf("failed to delete space %q (pre-test): %s", storeName, err)
+			}
+			if err := s3Cl.ensureSpace(storeName); err != nil {
+				t.Fatalf("failed to ensure space %q: %s", storeName, err)
+			}
+			defer func() {
+				if err := s3Cl.deleteSpace(storeName); err != nil {
+					t.Fatalf("failed to delete space %q (post-test): %s", storeName, err)
+				}
+			}()
+
+			// Create cluster.
+			extraEnvs := []string{
+				fmt.Sprintf("%s=do://%s", kopsEnvVarStateStore, storeName),
+				"KUBECONFIG=" + kubeConfFile,
+			}
+			if err := runScript(extraEnvs, "destroy_cluster.sh"); err != nil {
+				t.Fatalf("failed to destroy cluster (pre-test): %s", err)
+			}
+			if err := runScript(extraEnvs, "setup_cluster.sh", tt.kubeVer, strconv.Itoa(numWantNodes)); err != nil {
+				t.Fatalf("failed to set up cluster: %s", err)
+			}
+			defer func() {
+				if err := runScript(extraEnvs, "destroy_cluster.sh"); err != nil {
+					t.Errorf("failed to destroy cluster (post-test): %s", err)
+				}
+			}()
+
+			cs, err := kubeClient(kubeConfFile)
+			if err != nil {
+				t.Fatalf("failed to create Kubernetes client: %s", err)
+			}
+
+			// Check that nodes become ready.
+			l.Println("Polling for node readiness")
+			var (
+				gotNodes      []corev1.Node
+				numReadyNodes int
+			)
+			start := time.Now()
+			if err := wait.Poll(5*time.Second, 6*time.Minute, func() (bool, error) {
+				nl, err := cs.Core().Nodes().List(metav1.ListOptions{LabelSelector: "kubernetes.io/role=node"})
+				if err != nil {
+					return false, err
+				}
+
+				gotNodes = nl.Items
+				numReadyNodes = 0
+				for _, node := range gotNodes {
+					for _, cond := range node.Status.Conditions {
+						if cond.Type == corev1.NodeReady && cond.Status == corev1.ConditionTrue {
+							if _, ok := node.Labels[doLabel]; ok {
+								numReadyNodes++
+							}
+						}
+					}
+				}
+
+				l.Printf("Found %d/%d ready node(s)", numReadyNodes, numWantNodes)
+				return numReadyNodes == numWantNodes, nil
+			}); err != nil {
+				t.Fatalf("got %d ready node(s), want %d: %s\nnnodes: %v", numReadyNodes, numWantNodes, err, spew.Sdump(gotNodes))
+			}
+			l.Printf("Took %v for nodes to become ready\n", time.Since(start))
+
+			// Check that load balancer is working.
+
+			// Install example pod to load-balance to.
+			appName := "app"
```

Comment:
Probably good to append these names with some random hash to avoid conflicts in future runs (in case they're not cleaned up in previous runs).

Reply:
Right now, we remove the entire cluster after each Kubernetes version-specific sub-test run and create a brand new cluster for each sub-test, so it shouldn't be possible to run into a conflict situation. Or did I miss something?

Reply:
No I think it's fine then :)